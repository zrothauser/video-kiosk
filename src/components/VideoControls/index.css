.b-video-controls {
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: 100vw;
  height: 5.555vh;
  margin-bottom: 4.629vh;
  padding: 0 2%;
  z-index: 3;
  font-size: 20px;
  font-weight: 700;
}

.b-video-controls::after {
  display: block;
  content: '';
  width: 100vw;
  height: 10.184vh;
  position: absolute;
  top: 0;
  left: 0;
  z-index: -1;
  background: linear-gradient(to bottom, rgba(20, 20, 20, 0)0%, rgba(20, 20, 20, 1)100%);
}

/* Upper section */
.b-video-controls__upper {
  width: 100%;
  height: 11px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

/* Lower section */
.b-video-controls__lower {
  width: 100%;
  height: 2.85vh;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
}

.b-video-controls__title__category {
  text-transform: uppercase;
  letter-spacing: 0.16em;
  font-weight: 700;
}

/** Pagination */
.b-video-controls__pagination {
  margin-right: 1em;
  letter-spacing: .16em;
  color: rgba(255, 255, 255, .5);
}

.b-video-controls__pagination__numbers {
  margin: 0 0.5em;
}

.b-video-controls__pagination__current-index {
  padding-left: 0.1em;
  color: rgba(255, 255, 255, 1);
}

.b-video-controls__pagination__total {
  padding-left: 0.1em;
}

.b-video-controls__pagination__button {
  display: inline-block;
  height: 20px;
  width: 10px;
}

/* Controls */
.b-video-controls__topic-link {
  letter-spacing: 0.1em;
  font-size: 18px;
}

.b-video-controls__topic-icon {
  vertical-align: top;
  width: 1em;
  height: 1em;
  margin-right: 0.5em;
}

.b-video-controls__lower__right {
  display: flex;
  flex-wrap: nowrap;
  justify-content: flex-end;
  align-items: center;
}

.b-video-controls__lower__right__buttons {
  display: flex;
  font-size: 1.5em;
  letter-spacing: .5em;
  cursor: pointer;
  margin-left: 5.46vw;
  align-items: center;
  justify-content: space-between;
}

.b-video-controls__volume-button {
  padding: 0;
  -webkit-appearance: none;
  background: transparent;
  border: 0;
}

.b-video-controls__volume-icon {
  width: 2.1314vw;
  height: 2.85vh;
}

.b-video-controls__cc-button {
  margin-left: 1.562vw;
  padding: 0;
  -webkit-appearance: none;
  background: transparent;
  border: 0;
}

.b-video-controls__cc-icon {
  height: 2.85vh;
  width: 2.5vw;
}


/* Custom Sliders */
/*input[type=range] {
  opacity: 0.01;
  width: 100%;
  height: 100%;
  -webkit-appearance: none;
  margin: 0px;
  padding: 0px;
  border: 0 none;
  background: transparent;
  color: transparent;
  overflow: visible;
  z-index: 101;
}

input[type=range]:focus {
  outline: none;
}

input[type=range],
::-moz-range-track,
::-ms-track {
  width: 100%;
  height: 100%;
  -moz-appearance: none;
  margin: 0px;
  padding: 0px;
  border: 0 none;
  background: transparent;
  color: transparent;
  overflow: visible;
}

input[type=range]:focus::-webkit-slider-runnable-track {
  background: transparent;
  border: transparent;
}

input[type=range]::-webkit-slider-thumb,
::-moz-range-thumb,
::-ms-thumb {
  -webkit-appearance: none;
  outline: none;
  background-color: transparent;
  width: 21px;
  height: 21px;
}

input[type=range]::-ms-fill-lower {
  background: transparent;
  border: 0 none;
}

input[type=range]::-ms-fill-upper {
  background: transparent;
  border: 0 none;
}

input[type=range]::-ms-tooltip {
  display: none;
}

.volume-container {
  border: 0 none;
  overflow: visible;
  height: 21px;
  width: 100px;
  padding-left: 15px;
  margin-right: 1.562vw;
}

.scrubber-container {
  border: 0 none;
  position: absolute;
  overflow: visible;
  height: 21px;
  width: 100%;
}

.volume-slider,
.scrubber-slider {
  border: 0 none;
  position: relative;
  overflow: visible;
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
}

.slider {
  position: absolute;
  overflow: visible;
  z-index: 100;
}

.slidertrack {
  border-radius: 1px;
  position: absolute;
  height: 3px;
  width: 100%;
  background-color: rgba(255, 255, 255, .5);
}

.slidertrack-played {
  border-radius: 1px;
  position: absolute;
  height: 3px;
  left: 0;
  background-color: white;
}

.sliderthumb {
  width: 21px;
  height: 21px;
  border-radius: 50%;
  background-size: 21px 21px;
  background-position: 0px 0px;
  background-color: white;
  position: absolute;
  padding: 0;
  margin: 0;
  text-align: center;
  pointer-events: none;
}*/

/* Icons */
.current-time {
  display: flex;
  width: 100%;
  justify-content: flex-end;
  margin-top: -3vw;
}
