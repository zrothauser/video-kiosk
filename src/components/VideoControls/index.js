// Dependencies
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { AllHtmlEntities as Entities } from 'html-entities';
import Transition from 'react-transition-group/Transition';

// Helpers
import {
  getCategoryURL,
  getVideoURL,
} from '../../utils/navigation';

import { convertSecondsToMinutesSeconds } from '../../utils/video';

// Components
import Slider from '../Slider';

// Icons
import CCSVG from '../../resources/icons/cc.svg';
import VolumeSVG from '../../resources/icons/volume.svg';
import PreviousSVG from '../../resources/icons/notch-left.svg';
import NextSVG from '../../resources/icons/notch-right.svg';
import UpSVG from '../../resources/icons/notch-up.svg';

// Styles
import './index.css';

// Transitions
import * as transitions from '../transitions';

const {
  durations,
  shortFadeStyles,
  mediumFadeStyles,
} = transitions;

// Set up helper object
const entities = new Entities();

class VideoControls extends React.Component {
  /**
   * Renders the pagination controls part of the Controls.
   *
   * Includes previous/next buttons (if applicable), and the index
   * of the current video out of the total number of videos in the
   * category.
   */
  renderPaginationControls() {
    const {
      allVideosInCategory: allVideos,
      indexInCategory: index,
      parentCategory,
    } = this.props;

    // Don't render anything if there aren't any other videos
    if (allVideos.length === 0) {
      return null;
    }

    // Find the previous and next videos, if they exist
    const previousVideo = (index !== 0) ? allVideos[index - 1] : null;
    const nextVideo = (index + 1 < allVideos.length) ? allVideos[index + 1] : null;

    return (
      <span className="b-video-controls__pagination">
        {previousVideo &&
          <Link
            className="b-video-controls__pagination__button"
            to={getVideoURL(parentCategory, previousVideo)}
            rel="prev"
          >
            <img
              src={PreviousSVG}
              alt="Previous Video"
            />
          </Link>
        }

        <span className="b-video-controls__pagination__numbers">
          <span className="b-video-controls__pagination__current-index">
            {index + 1}
          </span>
          /
          <span className="b-video-controls__pagination__total">
            {allVideos.length}
          </span>
        </span>

        {nextVideo &&
          <Link
            className="b-video-controls__pagination__button"
            to={getVideoURL(parentCategory, nextVideo)}
            rel="next"
          >
            <img
              src={NextSVG}
              alt="Next Video"
            />
          </Link>
        }
      </span>
    );
  }

  render() {
    const {
      title,
      parentCategory,
      parentCategoryTitle,
      duration,
      currentTime,
      hasCaptions,
      volume,
      showVolumeControl,
      handleSeek,
      handleVolumeChange,
      toggleVolumeControl,
      toggleCaptions,
      visible,
    } = this.props;

    return (
      <Transition
        in={visible}
        timeout={durations.medium}
        appear
      >
        {state => (
          <div
            className="b-video-controls"
            style={{
              ...mediumFadeStyles.default,
              ...mediumFadeStyles[state],
            }}
          >
            <div className="b-video-controls__upper">
              <div className="b-video-controls__time">
                {convertSecondsToMinutesSeconds(currentTime)}
              </div>

              <Slider
                minimum={0}
                maximum={duration}
                value={currentTime}
                handleSeek={handleSeek}
              />
            </div>

            <div className="b-video-controls__lower">
              {this.renderPaginationControls()}

              <span className="b-video-controls__title">
                <span className="b-video-controls__title__category">
                  {parentCategoryTitle}:
                </span>
                {entities.decode(title)}
              </span>

              <span className="b-video-controls__lower__right">
                <Link
                  className="b-video-controls__topic-link"
                  to={getCategoryURL(parentCategory)}
                >
                  <img
                    src={UpSVG}
                    alt="Closed Captions Button"
                    className="b-video-controls__topic-icon"
                  />
                  BACK TO TOPIC
                </Link>

                <span className="b-video-controls__lower__right__buttons">
                  <Transition
                    in={showVolumeControl}
                    timeout={durations.shortest}
                  >
                    {controlState => (
                      <div
                        className="b-video-controls__volume-slider"
                        style={{
                          ...shortFadeStyles.default,
                          ...shortFadeStyles[controlState],
                        }}
                      >
                        <Slider
                          minimum={0}
                          maximum={100}
                          value={volume}
                          handleSeek={handleVolumeChange}
                        />
                      </div>
                    )}
                  </Transition>

                  <button
                    className="b-video-controls__volume-button"
                    onClick={() => toggleVolumeControl()}
                  >
                    <img
                      src={VolumeSVG}
                      alt="Volume Controls"
                      className="b-video-controls__volume-icon"
                    />
                  </button>

                  {hasCaptions &&
                    <button
                      className="b-video-controls__cc-button"
                      onClick={() => toggleCaptions()}
                    >
                      <img
                        src={CCSVG}
                        alt="Closed Captions Button"
                        className="b-video-controls__cc-icon"
                      />
                    </button>
                  }
                </span>
              </span>
            </div>

            <span className="b-video-controls__scrim" />
          </div>
        )}
      </Transition>
    );
  }
}

VideoControls.propTypes = {
  title: PropTypes.string.isRequired,
  parentCategory: PropTypes.string.isRequired,
  parentCategoryTitle: PropTypes.string.isRequired,
  duration: PropTypes.number.isRequired,
  currentTime: PropTypes.number.isRequired,
  allVideosInCategory: PropTypes.arrayOf(PropTypes.string).isRequired,
  indexInCategory: PropTypes.number.isRequired,
  hasCaptions: PropTypes.bool.isRequired,
  volume: PropTypes.number.isRequired,
  showVolumeControl: PropTypes.bool.isRequired,
  handleSeek: PropTypes.func.isRequired,
  handleVolumeChange: PropTypes.func.isRequired,
  toggleVolumeControl: PropTypes.func.isRequired,
  toggleCaptions: PropTypes.func.isRequired,
  visible: PropTypes.bool.isRequired,
};

export default VideoControls;
