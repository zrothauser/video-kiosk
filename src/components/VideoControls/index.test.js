import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';

import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import VideoControls from './index';

describe('components', () => {
  describe('VideoControls', () => {
    const mockProps = {
      title: 'Test Title',
      parentCategory: 'go-deeper',
      parentCategoryTitle: 'Go Deeper',
      duration: 122,
      currentTime: 30,
      allVideosInCategory: [],
      indexInCategory: 0,
      hasCaptions: true,
      volume: 60,
      handleSeek: jest.fn(),
      handleVolumeChange: jest.fn(),
      toggleCaptions: jest.fn(),
      visible: true,
      areCaptionsVisible: true,
    };

    it('renders without crashing', () => {
      shallow(<VideoControls {...mockProps} />);
    });

    it('renders correctly', () => {
      const VideoControlsComponent = renderer.create(
        <Router>
          <VideoControls {...mockProps} />
        </Router>,
      );

      expect(VideoControlsComponent).toMatchSnapshot();
    });
  });
});
