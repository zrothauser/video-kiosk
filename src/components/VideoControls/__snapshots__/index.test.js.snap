// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components VideoControls renders correctly 1`] = `
<div
  className="b-video-controls"
  style={
    Object {
      "opacity": 0,
      "transition": "opacity 400ms ease-in-out",
    }
  }
>
  <div
    className="b-video-controls__upper"
  >
    <div
      className="b-video-controls__time"
    >
      0:30
    </div>
    <div
      className="b-slider"
    >
      <div
        className="b-slider__track"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="b-slider__progress"
          style={
            Object {
              "width": "24.59016393442623%",
            }
          }
        />
      </div>
      <div
        aria-valuemax={122}
        aria-valuemin={0}
        aria-valuenow={30}
        className="b-slider__thumb"
        role="slider"
        style={
          Object {
            "left": "24.59016393442623%",
          }
        }
        tabIndex="0"
      />
    </div>
  </div>
  <div
    className="b-video-controls__lower"
  >
    <span
      className="b-video-controls__title"
    >
      <span
        className="b-video-controls__title__category"
      >
        Go Deeper
        :
      </span>
      Test Title
    </span>
    <span
      className="b-video-controls__lower__right"
    >
      <a
        className="b-video-controls__topic-link"
        href="//go-deeper/"
        onClick={[Function]}
      >
        <img
          alt="Closed Captions Button"
          className="b-video-controls__topic-icon"
          src="notch-up.svg"
        />
        BACK TO TOPIC
      </a>
      <span
        className="b-video-controls__lower__right__buttons"
      >
        <button
          className="b-video-controls__volume-button"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="Volume Controls"
            className="b-video-controls__volume-icon"
            src="volume.svg"
          />
        </button>
        <button
          className="b-video-controls__cc-button"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="Closed Captions Button"
            className="b-video-controls__cc-icon"
            src="captions-on.svg"
          />
        </button>
      </span>
    </span>
  </div>
  <span
    className="b-video-controls__scrim"
  />
</div>
`;
